### PR TITLE
[ASPagerNode] Add ASPagerNode rotation support

### LINKED ISSUE #1169: [ASPagerNode] Add support for rotation.
I've swapped out a `UIPageViewController` and replaced it with an `ASPagerNode`. Works great, but when rotating/transitioning to a new size, the pager node doesn't remain on the current page, and instead scrolls off to the left/right and leaves the ASPagerNode halfway between pages.

Also receive the following error:
```
2016-02-03 14:53:34.323 AFR-dev[1227:400873] the behavior of the UICollectionViewFlowLayout is not defined because:
2016-02-03 14:53:34.323 AFR-dev[1227:400873] the item height must be less than the height of the UICollectionView minus the section insets top and bottom values, minus the content insets top and bottom values.
2016-02-03 14:53:34.326 AFR-dev[1227:400873] The relevant UICollectionViewFlowLayout instance is <UICollectionViewFlowLayout: 0x141096330>, and it is attached to <ASCollectionView: 0x1400e2e00; baseClass = UICollectionView; frame = (0 0; 1024 768); clipsToBounds = YES; gestureRecognizers = <NSArray: 0x14105a090>; animations = { position=<CABasicAnimation: 0x141080490>; bounds=<CABasicAnimation: 0x14162a6f0>; }; layer = <<ASPagerNode: 0x13ff97d00>, layer = <_ASDisplayLayer: 0x13ffa3210>>; contentOffset: {0, 0}; contentSize: {22272, 1024}> collection view layout: <UICollectionViewFlowLayout: 0x141096330>.
2016-02-03 14:53:34.326 AFR-dev[1227:400873] Make a symbolic breakpoint at UICollectionViewFlowLayoutBreakForInvalidSizes to catch this in the debugger.
```

### PR DESCRIPTION
**Pull request by maicki**
Build on top of #1224 and should fix #1169

### CODE REVIEW

**Comment by maicki** [comment]
@levi @appleguy Can I get a code review please

**Comment by appleguy** [comment]
cc @levi, let me know if you can't fit in this review on Monday - I'd also be personally interested to learn more about the code / challenge at hand to provide a good review, though also have other reviews to dig into.

File: AsyncDisplayKit/ASPagerFlowLayout.m [hunk 1]
```diff
@@ -14,58 +14,71 @@ @interface ASPagerFlowLayout ()
 
 @end
 
-@implementation ASPagerFlowLayout
+@implementation ASPagerFlowLayout {
```
Comment (levi) [inline AsyncDisplayKit/ASPagerFlowLayout.m]: Add these ivars to a category extension of ASPagerFlowLayout.

File: AsyncDisplayKit/ASPagerNode.h [hunk 2]
```diff
@@ -84,7 +84,10 @@
 // The underlying ASCollectionView object.
 @property (nonatomic, readonly) ASCollectionView *view;
 
+/// Returns the current page index
+@property (nonatomic, assign, readonly) NSInteger currentPageIndex;
+
+/// Scrolls the pager node contents until the specified page is visible.
```
Comment (levi) [inline AsyncDisplayKit/ASPagerNode.h]: Use Java-doc style:
/**
 *
 */

File: AsyncDisplayKit/ASPagerFlowLayout.m [hunk 3]
```diff
@@ -14,58 +14,71 @@ @interface ASPagerFlowLayout ()
...
-  NSArray<UICollectionViewLayoutAttributes *> *layoutAttributes = [self layoutAttributesForElementsInRect:visibleRect];
-  for (UICollectionViewLayoutAttributes *attributes in layoutAttributes) {
-    if ([attributes representedElementCategory] == UICollectionElementCategoryCell && attributes.center.x == visibleXCenter) {
-      return attributes.indexPath;
-    }
-  }
-  return nil;
+  
+  CGFloat currentPageXOffset = self.currentIndexPath.item * CGRectGetWidth(self.collectionView.bounds);
+    return CGPointMake(currentPageXOffset, proposedContentOffset.y);
```
Comment (levi) [inline AsyncDisplayKit/ASPagerFlowLayout.m]: indentation

File: AsyncDisplayKit/ASPagerFlowLayout.m [hunk 4]
```diff
@@ -14,58 +14,71 @@ @interface ASPagerFlowLayout ()
 
 @end
 
-@implementation ASPagerFlowLayout
+@implementation ASPagerFlowLayout {
+    BOOL _didRotate;
+    CGRect _cachedCollectionViewBounds;
```
Comment (levi) [inline AsyncDisplayKit/ASPagerFlowLayout.m]: Be sure to use 2 spaces.

**Comment by levi** [comment]
@maicki tested this on the Pinterest app. The result looks ok, but the rotation animation is busted. I'd recommend pulling it down locally and rotating while you're in pin closeup — that's the most common case.

**Comment by appleguy** [comment]
@maicki @levi is this now tested to work with a smooth animation in Pinterest?  Ready to land?

**Comment by maicki** [comment]
@appleguy No I haven't had time yet to work on it

**Comment by appleguy** [comment]
@maicki just to be clear, do you mean you haven't had a chance to check it, or you know it still needs work / should not land?  I can check it if we don't know if your latest changes fixed it.

**Comment by maicki** [comment]
@appleguy I haven't had a chance to check / fix it after my last change and the comment from @levi

**Comment by maicki** [comment]
@appleguy @levi Ready for review again. Also in the Pinterest app. Rotation animation resembles the animation before, but should fix #1169

File: examples/PagerNode/Sample/ViewController.m [hunk 5]
```diff
@@ -47,11 +63,13 @@ - (NSInteger)numberOfPagesInPagerNode:(ASPagerNode *)pagerNode
   return 5;
 }
 
-- (ASCellNode *)pagerNode:(ASPagerNode *)pagerNode nodeAtIndex:(NSInteger)index
+- (ASCellNodeBlock)pagerNode:(ASPagerNode *)pagerNode nodeBlockAtIndex:(NSInteger)index
 {
-  PageNode *page = [[PageNode alloc] init];
-  page.backgroundColor = [UIColor blueColor];
-  return page;
```
Comment (appleguy) [inline examples/PagerNode/Sample/ViewController.m]: For sample code, we should add a comment above the block saying something like (feel free to edit or reduce):

// NOTE: For implementers of this method, it is critical to use the value of "index" outside of the block to access any data source objects, etc.
// The block is executed concurrently in the background and data source values may change at a given index.

**Comment by appleguy** [comment]
@maicki this code is especially clean!  Great, succinct implementation.  Thanks for updating the example code.  It's not common now, but soon we should add more comments to example code for readers to learn the maximum.

**Pull request merged**